src/index.rs

```diff
@@ -1,132 +1,278 @@
 /// Construct an Index.
-use crate::prelude::{State, TransitionKey};
-use crate::regex::{get_vocabulary_transition_keys, state_scan_tokens};
+use crate::prelude::*;
 use crate::vocabulary::Vocabulary;
 use crate::{Error, Result};
 use bincode::{Decode, Encode};
-use rustc_hash::{FxHashMap, FxHashSet};
-
-#[derive(Debug)]
-pub struct FSMInfo {
-    pub(crate) initial: State,
-    pub(crate) finals: FxHashSet<State>,
-    pub(crate) transitions: FxHashMap<(State, TransitionKey), State>,
-    pub(crate) alphabet_anything_value: TransitionKey,
-    pub(crate) alphabet_symbol_mapping: FxHashMap<String, TransitionKey>,
-}
-
-impl FSMInfo {
-    pub fn new(
-        initial: State,
-        finals: FxHashSet<State>,
-        transitions: FxHashMap<(State, TransitionKey), State>,
-        alphabet_anything_value: TransitionKey,
-        alphabet_symbol_mapping: FxHashMap<String, TransitionKey>,
-    ) -> Self {
-        Self {
-            initial,
-            finals,
-            transitions,
-            alphabet_anything_value,
-            alphabet_symbol_mapping,
-        }
-    }
-}
+use regex_automata::dfa::{dense::DFA, Automaton};
+use regex_automata::util::primitives::StateID as AutomataStateId;
+use regex_automata::Anchored;
+use rustc_hash::{FxHashMap as HashMap, FxHashSet as HashSet};
 
-#[derive(Debug, Encode, Decode)]
+#[derive(Clone, Debug, PartialEq, Encode, Decode)]
 pub struct Index {
-    initial: u32,
-    finals: FxHashSet<u32>,
-    states_to_token_subsets: FxHashMap<u32, FxHashMap<u32, u32>>,
-    eos_token_id: u32,
+    /// The ID of the initial state in the automaton, processing begins from this state.
+    initial_state: StateId,
+    /// A collection of states considered as terminal states.
+    final_states: HashSet<StateId>,
+    /// A mapping of state transitions, defined by tokens ids and their corresponding state changes:
+    ///   - The outer map's keys are the state IDs.
+    ///   - The inner map's keys are token IDs.
+    ///   - The inner map's values are state IDs, indicating transitions to the next state.
```

Comment:
These are already in the type signature. I think an example transitions map and it's explanation would be better here.

Reply:
Let me think of something more useful here

Reply:
Tried with words, gets confusing fast, so ended up with this 🖼️ 
![image](https://github.com/user-attachments/assets/a25972e2-d1bf-460d-b3d8-19b9f4eb92c5)